processor/batch-processor/src/run.ts

```diff
@@ -110,66 +137,120 @@ class Processor<B extends BlockBase, S> {
+        let chainHeight = await this.chainHeight.get()
+
         this.metrics.setChainHeight(chainHeight)
         this.metrics.setLastProcessedBlock(state.height)
         let left: number
         let processed: number
         if (this.src.getBlocksCountInRange) {
             left = this.src.getBlocksCountInRange({
                 from: this.metrics.getLastProcessedBlock() + 1,
-                to: this.metrics.getChainHeight()
+                to: this.metrics.getChainHeight(),
             })
-            processed = this.src.getBlocksCountInRange({
-                from: 0,
-                to: this.metrics.getChainHeight()
-            }) - left
+            processed =
+                this.src.getBlocksCountInRange({
+                    from: 0,
+                    to: this.metrics.getChainHeight(),
+                }) - left
         } else {
             left = this.metrics.getChainHeight() - this.metrics.getLastProcessedBlock()
             processed = this.metrics.getLastProcessedBlock()
         }
         this.metrics.updateProgress(processed, left, time)
     }
 
-    private async processBatch(prevHead: HashAndHeight, blocks: B[]): Promise<HashAndHeight> {
-        let chainHeight = await this.chainHeight.get()
+    private async processBatch(prevState: HotDatabaseState, finalizedHead: HashAndHeight, blocks: B[]): Promise<HotDatabaseState> {
+        assert(
+            prevState.height < finalizedHead.height ||
+                (prevState.height === finalizedHead.height && prevState.hash === finalizedHead.hash)
+        )
+        assert(
+            prevState.height < blocks[0].header.height ||
+                (prevState.height === blocks[0].header.height && prevState.hash === blocks[0].header.hash)
+        )
 
-        let nextHead = {
-            hash: last(blocks).header.hash,
-            height: last(blocks).header.height
+        let lastBlock = last(blocks).header
+        let nextState: HotDatabaseState =
+            lastBlock.height < finalizedHead.height
+                ? toDatabaseState(lastBlock)
+                : toDatabaseState(finalizedHead)
+        
+        let baseHead = prevState as HashAndHeight
+        for (let block of prevState.top) {
+            if (block.height > blocks[0].header.height) break
+            if (block.height === blocks[0].header.height && block.hash !== blocks[0].header.hash) break
+            baseHead = block
+            if (block.height <= nextState.height) continue
+            nextState.top.push(block)
         }
 
-        let isOnTop = nextHead.height >= chainHeight
+        for (let {header: block} of blocks) {
+            if (block.height <= finalizedHead.height) continue
+            nextState.top.push(block)
+        }
 
-        let mappingStartTime = process.hrtime.bigint()
+        let prevHead = maybeLast(prevState.top) || prevState
+        let nextHead = maybeLast(nextState.top) || nextState
```

Comment:
State tracking procedure above feels tricky, but most importantly it is not need (both for the current  and future database interfaces).

```ts
let baseHead = getParent(blocks[0])
let nextHead = last(blocks)
```

Reply:
Processor doesn't know anything about parant blocks. Parant hash is not a part of the DataSource interface